### PR TITLE
[LSP] Don't generate semantic tokens ResultId unless necessary

### PR DESCRIPTION
**Pull request by allisonchou**
This should result in a small perf improvement. Today, a new ResultId is generated for each semantic tokens edit request (called every ~0.5s) regardless of whether the doc's content has changed. We can switch to re-using the old ResultId instead, which will avoid us inserting into the cache unnecessarily.

cc @ryanbrandenburg

### CODE REVIEW

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs [hunk 1]
```diff
@@ -56,24 +56,31 @@ public SemanticTokensEditsHandler(SemanticTokensCache tokensCache)
...
             var oldSemanticTokensData = await _tokensCache.GetCachedTokensDataAsync(
                 request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
             if (oldSemanticTokensData == null)
             {
-                return newSemanticTokens;
+                var newResultId = _tokensCache.GetNextResultId();
+                return new LSP.SemanticTokens { ResultId = newResultId, Data = newSemanticTokensData };
+            }
+
+            var resultId = request.PreviousResultId;
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:69]:
```suggestion
```

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs [hunk 2]
```diff
@@ -56,24 +56,31 @@ public SemanticTokensEditsHandler(SemanticTokensCache tokensCache)
...
+
+            var editArray = ComputeSemanticTokensEdits(oldSemanticTokensData, newSemanticTokensData);
+            if (editArray.Length != 0)
+            {
+                // If we have edits, generate a new ResultId. Otherwise, re-use the previous one.
+                resultId = _tokensCache.GetNextResultId();
+                var updatedTokens = new LSP.SemanticTokens { ResultId = resultId, Data = newSemanticTokensData };
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:76]:
```suggestion
                var updatedTokens = new LSP.SemanticTokens { ResultId = GetNextResultId(), Data = newSemanticTokensData };
```

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs [hunk 3]
```diff
@@ -56,24 +56,31 @@ public SemanticTokensEditsHandler(SemanticTokensCache tokensCache)
...
+                await _tokensCache.UpdateCacheAsync(
+                    request.TextDocument.Uri, updatedTokens, cancellationToken).ConfigureAwait(false);
             }
 
             var edits = new SemanticTokensEdits
             {
-                Edits = ComputeSemanticTokensEdits(oldSemanticTokensData, newSemanticTokensData),
+                Edits = editArray,
                 ResultId = resultId
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:84]:
```suggestion
                ResultId = request.PreviousResultId;
```

**Review by CyrusNajmabadi — DISMISSED** [review]
Request changes
Submit feedback that must be addressed before merging.

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs [hunk 4]
```diff
@@ -56,25 +56,29 @@ public SemanticTokensEditsHandler(SemanticTokensCache tokensCache)
...
             {
-                return newSemanticTokens;
+                var newResultId = _tokensCache.GetNextResultId();
+                return new LSP.SemanticTokens { ResultId = newResultId, Data = newSemanticTokensData };
+            }
+
+            var editArray = ComputeSemanticTokensEdits(oldSemanticTokensData, newSemanticTokensData);
+            if (editArray.Length != 0)
+            {
+                // If we have edits, generate a new ResultId. Otherwise, re-use the previous one.
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs:72]: comment is a little misplaced.  but nbd.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by allisonchou** [comment]
@CyrusNajmabadi I ended up reverting the suggestions because I didn't realize they actually change the code's semantics. :(
When we return `SemanticTokensEdits`, the associated ResultId should either be incremented if we return 1+ edit, or be equal to the previous ResultId if no edits are returned. I believe the provided suggestions made it so we always returned the previous ResultId.
Regardless, I really appreciate you taking the time to leave feedback!

**Pull request merged**